iocore/net/quic/QUICAckFrameCreator.cc

```diff
@@ -141,79 +87,178 @@ QUICAckFrameCreator::will_generate_frame(QUICEncryptionLevel level)
   }
 
   int index = QUICTypeUtil::pn_space_index(level);
-  return this->_should_send[index];
-}
-
-uint64_t
-QUICAckFrameCreator::_calculate_delay(QUICEncryptionLevel level)
-{
-  // Ack delay is in microseconds and scaled
-  ink_hrtime now             = Thread::get_hrtime();
-  int index                  = QUICTypeUtil::pn_space_index(level);
-  uint64_t delay             = (now - this->_packet_numbers[index].largest_ack_received_time()) / 1000;
-  uint8_t ack_delay_exponent = 3;
-  if (level != QUICEncryptionLevel::INITIAL && level != QUICEncryptionLevel::HANDSHAKE) {
-    ack_delay_exponent = this->_ack_delay_exponent;
-  }
-  return delay >> ack_delay_exponent;
+  return this->_packet_numbers[index].will_generate_frame(level);
 }
 
 //
 // QUICAckPacketNumbers
 //
 void
-QUICAckPacketNumbers::push_back(QUICPacketNumber packet_number)
+QUICAckFrameCreator::QUICAckPacketNumbers::push_back(QUICPacketNumber packet_number, size_t size, bool ack_only)
 {
   if (packet_number == 0 || packet_number > this->_largest_ack_number) {
     this->_largest_ack_received_time = Thread::get_hrtime();
     this->_largest_ack_number        = packet_number;
   }
 
-  this->_packet_numbers.push_back(packet_number);
+  // FIXME: fix size for second full-size
+  if (packet_number != this->_largest_ack_number + 1 || this->_latest_recv_size + size > 2 * 1480) {
+    this->_should_send = true;
+  }
+
+  if (!ack_only) {
+    this->_is_availiable |= !ack_only;
+  }
+
+  // TODO do we really need to send ack for every second full-sized packet
+  // PS: we do the same thing as tcp
+  // An acknowledgement MAY be sent for every second full-sized packet, as TCP does [RFC5681], or may be sent less frequently, as
+  // long as the delay does not exceed the maximum ack delay. QUIC recovery algorithms do not assume the peer generates an
+  // acknowledgement immediately when receiving a second full-sized packet.
+  this->_latest_recv_size += size;
```

Comment:
OK, I didn't read that part (I thought everything is on the pseudo code).

We definitely need to do something for this but I'm still not sure whether we should do this inside QUICAckPacketNumbers. I think QUICAckFrameCreator should be responsible for a timing of sending ACK, based on:
- Frames received
- Allowed delay
- Size of packet received

It would be great if we could clearly see these three as criteria.

Give me some more time to review.